feat(vertexai): Add a new sample for automatic function calling

jaycee-li opened the pull request:
## Description

This PR added a new sample `AutomaticFunctionCalling.java` for automatic function calling usage. It also updated the library version and removed a deprecated field in `GroundingWithPublicData.java`.

## Checklist

- [x] I have followed [Sample Format Guide](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md)
- [x] `pom.xml` parent set to latest `shared-configuration`
- [ ] Appropriate changes to README are included in PR
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] **Tests** pass:   `mvn clean verify` **required**
- [x] **Lint**  passes: `mvn -P lint checkstyle:check` **required**
- [ ] **Static Analysis**:  `mvn -P lint clean compile pmd:cpd-check spotbugs:check` **advisory only**
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample 
- [x] Please **merge** this PR for me once it is approved

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+package vertexai.gemini;
+
+// [START generativeaionvertexai_gemini_automatic_function_calling]
+// [START aiplatform_gemini_automatic_function_calling]ß
```
```suggestion
// [START aiplatform_gemini_automatic_function_calling]
```

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+
+    String promptText = "What's the weather like in Paris?";
+
+    whatsTheWeatherLike(projectId, location, modelName, promptText);
```
Make the method name generic.

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+    whatsTheWeatherLike(projectId, location, modelName, promptText);
+  }
+
+  /** Callable function getCurrentWeather. */
```
Can we add a comment clarifying this is just a stub and external functions with http calls can be substituted here?

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  // A request involving the interaction with an external tool
+  public static String whatsTheWeatherLike(
```
Update method name.

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+
+  // A request involving the interaction with an external tool
```
Update the description to provide an overview of function calling.

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+      FunctionDeclaration functionDeclaration =
+          FunctionDeclarationMaker.fromFunc(
+              "Get the current weather in a given location", function, "location");
+      System.out.println("Function declaration:");
+      System.out.println(functionDeclaration);
```
```suggestion
      System.out.printf("Function declaration: %s", functionDeclaration);
```

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+          chat.withAutomaticFunctionCallingResponder(responder).sendMessage(promptText);
+
+      // Check the final response
+      System.out.println("Print response: ");
```
```suggestion
```

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+      // Check the final response
+      System.out.println("Print response: ");
+      String finalAnswer = ResponseHandler.getText(response);
+      System.out.println(finalAnswer);
```
```suggestion
      System.out.printf"Response: %s", finalAnswer);
```

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -0,0 +1,99 @@
+      // Send message with the responder, which auto-responds FunctionCalls and
+      // returns the final text result
+      GenerateContentResponse response =
+          chat.withAutomaticFunctionCallingResponder(responder).sendMessage(promptText);
```
If the `withAutomaticFunctionCallingResponder(responder)` is not chained here, will the function be auto-inferred? What's the use of chaining the function here when it's already present as part of the tool?

jaycee-li replied:
No, the function won't be auto-inferred if `withAutomaticFunctionCallingResponder(responder)` is not presented.

While adding the tool introduces the "Function Calling" feature, enabling the model to potentially respond with a `FunctionCall` instruction, users still need to manually execute the function, construct a `FunctionResponse`, and make another request.

Adding the responder activates "Automatic Function Calling" feature, automating these manual steps within the client library, saving users time and code. However, as not all functions are callable within the same Java runtime, this feature is disabled by default, users need to explicit chain it when making the request.

Sita04 replied:
Thanks for the clarification!

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/GroundingWithPublicData.java:
```diff
@@ -43,15 +43,16 @@ public static String groundWithPublicData(String projectId, String location, Str
-      Tool googleSearchTool = Tool.newBuilder()
-          .setGoogleSearchRetrieval(
-              // Enable using the result from this tool in detecting grounding
-              GoogleSearchRetrieval.newBuilder().setDisableAttribution(false))
```
Why are we turning off the attribution here?

jaycee-li replied:
This is a preview attribution for testers, and the backend has removed this field.

Sita04 commented:
Kindly fix the test failure.

Sita04 commented:
@jaycee-li gentle ping on this PR.

jaycee-li commented:
> @jaycee-li gentle ping on this PR.

Hi @Sita04 , apologies for the delay, I was OOO for the past 4 weeks.

Thanks for the review! I've addressed all the comments.

Sita04 commented on vertexai/snippets/src/main/java/vertexai/gemini/AutomaticFunctionCalling.java:
```diff
@@ -52,8 +54,8 @@ public static String getCurrentWeather(String location) {
 
-  // A request involving the interaction with an external tool
-  public static String whatsTheWeatherLike(
+  // Use the Automatic Function Calling feature to auto-response Function Call requests from the model.
```
```suggestion
  // Use the Automatic Function Calling feature to auto-respond to Function Call requests from the model.
```

Sita04 reviewed (approved):
One nit. LGTM

jaycee-li commented:
Hi @Sita04, the previous integration test failed and I've fixed the issue. Could you please re-run the test? Thank you!
